File: 2025/python/p04.py
from typing import Generator


def coord(i: int, j: int) -> complex:
    return j + i * 1j


def neighbors(coord: complex) -> Generator[complex, None, None]:
    yield coord + 1
    yield coord - 1
    yield coord + 1j
    yield coord - 1j
    yield coord + 1 + 1j
    yield coord - 1 + 1j
    yield coord + 1 - 1j
    yield coord - 1 - 1j


def process(s) -> set[complex]:
    rolls = set()
    for i, line in enumerate(s):
        for j, x in enumerate(line):
            if x == "@":
                rolls.add(coord(i, j))
    return rolls


list(neighbors(1))

test_string = """..@@.@@@@.
@@@.@.@.@@
@@@@@.@.@@
@.@@@@..@.
@@.@@@@.@@
.@@@@@@@.@
.@.@.@.@@@
@.@@@.@@@@
.@@@@@@@@.
@.@.@@@.@."""

with open("../input/04.txt") as f:
    data = process(f)

test_data = process(test_string.splitlines())


def reachable(rolls: set[complex]) -> set[complex]:
    can_reach = set()
    for roll in rolls:
        if len(set(neighbors(roll)) & rolls) < 4:
            can_reach.add(roll)
    return can_reach


reachable(test_data)


def part1(data: set[complex]) -> int:
    return len(reachable(data))


assert part1(test_data) == 13, "Failed part 1 test!"


def part2(data: set[complex]) -> int:
    total = 0
    while can_reach := reachable(data):
        total += len(can_reach)
        data -= can_reach
    return total


assert part2(test_data) == 43, "Failed part 2 test!"


if __name__ == "__main__":
    ans1 = part1(data)
    print(f"Answer 1: {ans1}")

    ans2 = part2(data)
    print(f"Answer 2: {ans2}")

File: 2025/js/p04.ts
import { expect } from "bun:test";

const path = "../input/04.txt";
const file = Bun.file(path);
const dataString = await file.text();

const testString = `..@@.@@@@.
@@@.@.@.@@
@@@@@.@.@@
@.@@@@..@.
@@.@@@@.@@
.@@@@@@@.@
.@.@.@.@@@
@.@@@.@@@@
.@@@@@@@@.
@.@.@@@.@.`;

type Position = [number, number];
type Rolls = Map<string, Position>;

function coord(row: number, col: number): string {
  return `${row},${col}`;
}

function process(inp: string): Rolls {
  const rolls: Rolls = new Map();
  inp.split("\n").forEach((line, row) => {
    [...line].forEach((chr, col) => {
      if (chr === "@") {
        rolls.set(coord(row, col), [row, col]);
      }
    });
  });
  return rolls;
}

function* neighbors([row, col]: Position): Generator<string> {
  yield coord(row + 1, col);
  yield coord(row - 1, col);
  yield coord(row, col + 1);
  yield coord(row, col - 1);
  yield coord(row + 1, col + 1);
  yield coord(row - 1, col + 1);
  yield coord(row + 1, col - 1);
  yield coord(row - 1, col - 1);
}

const data = process(dataString);
const testData = process(testString);

function reachable(data: Rolls): Rolls {
  const canReach: Rolls = new Map();
  for (const [key, pos] of data) {
    let neighborCount = 0;
    for (const nkey of neighbors(pos)) {
      if (data.has(nkey)) neighborCount++;
    }
    if (neighborCount < 4) {
      canReach.set(key, pos);
    }
  }
  return canReach;
}

function part1(data: Rolls): number {
  return reachable(data).size;
}

expect(part1(testData)).toEqual(13);

const ans1 = part1(data);
console.log(`Answer 1: ${ans1}`);

function part2(input: Rolls): number {
  const data: Rolls = new Map(input);
  let canReach;
  let total = 0;
  while ((canReach = reachable(data)).size) {
    total += canReach.size;
    for (const key of canReach.keys()) {
      data.delete(key);
    }
  }
  return total;
}

expect(part2(testData)).toEqual(43);

const ans2 = part2(data);
console.log(`Answer 2: ${ans2}`);
